src/lib.rs

```diff
@@ -40,19 +42,314 @@ mod repl_helper;
+        /// [`scryer_start_new_query_generator`] that has not yet been freed with 
+        /// [`scryer_cleanup_query_generator`] or the [`Machine`] state will enter an undefined 
+        /// configuration with unpredictable results.  
+        ///
+        /// Other concerns:
+        /// * after invoking this function, calling any other function besides [`scryer_run_query_generator`]
+        /// before invoking [`scryer_cleanup_query_generator`] on the [`Machine`] pointer and [`QueryState`] pointer
+        /// will leave the [`Machine`] in an undefined state
+        ///
+        /// # Arguments
+        ///
+        /// * `machine` - A mutable reference to a [`Machine`] object.
+        /// * `input` - A pointer to a null terminated UTF-8 string.
+        ///
+        /// # Returns
+        ///
+        /// If the input string is successfully converted to a Rust string, the function returns a mutable pointer to a [`QueryState`] object.
+        /// If the input string is not a valid UTF-8 string or an error occurs during the query generation, the function returns a null pointer.
+        ///
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_start_new_query_generator(machine: &mut Machine, input: *const c_char) -> *mut QueryState {
+            match unsafe { CStr::from_ptr(input) }.to_str() {
+                Ok(input) => {
+                    let query_state = machine.start_new_query_generator(input.to_string());
+                    Box::into_raw(Box::new(query_state))
+                }
+                Err(err) => {
+                    eprintln!("Query input wasn't valid UTF-8: {err:?}");
+                    std::ptr::null_mut()
+                }
+            }
+        }
+
+
+        /// Cleans up the [`QueryState`] in the associated Scryer [`Machine`].
+        ///
+        /// # Safety
+        ///
+        /// * `machine` must be a valid mutable pointer to a [`Machine`] instance created by [`scryer_machine_new`] that has not yet been freed 
+        /// with [`scryer_machine_free`].
+        /// * `query_state` must be a valid mutable pointer to a [`QueryState`] created by [`scryer_start_new_query_generator`] 
+        /// that has not yet been freed by [`scryer_cleanup_query_generator`].
+        /// * There must be no other [`QueryState`] for this [`Machine`] started by 
+        /// [`scryer_start_new_query_generator`] that has not yet been freed with 
+        /// [`scryer_cleanup_query_generator`] or the [`Machine`] state will enter an undefined 
+        /// configuration with unpredictable results.  
+        ///
+        /// # Arguments
+        ///
+        /// - `machine`: A mutable pointer to the [`Machine`] instance.
+        /// - `query_state`: A mutable pointer to the [`QueryState`] instance.
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_cleanup_query_generator(machine: *mut Machine, query_state: *mut QueryState) {
```

Comment:
This would be the equivalent of the `Drop` implementation of `QueryState`.

```suggestion
        pub unsafe extern "C" fn scryer_query_state_free(machine: *mut Machine, query_state: *mut QueryState) {
```